File: android/app/src/main/java/com/dismoi/scout/Floating/Layout/Coordinator.java
package com.dismoi.scout.Floating.Layout;

import android.view.View;
import android.view.WindowManager;

import com.dismoi.scout.Floating.Layout.Trash;
import com.dismoi.scout.Floating.FloatingService;

final public class Coordinator {
  private static Coordinator INSTANCE;
  private Trash trashView;
  private WindowManager windowManager;
  private FloatingService bubblesService;

  private static Coordinator getInstance() {
    if (INSTANCE == null) {
      INSTANCE = new Coordinator();
    }
    return INSTANCE;
  }

  private Coordinator() { }

  public void notifyBubblePositionChanged(Bubble bubble, int x, int y) {
    if (trashView != null) {
      trashView.setVisibility(View.VISIBLE);
      if (checkIfBubbleIsOverTrash(bubble)) {
        trashView.applyMagnetism();
        trashView.vibrate();
        applyTrashMagnetismToBubble(bubble);
      } else {
        trashView.releaseMagnetism();
      }
    }
  }

  private void applyTrashMagnetismToBubble(Bubble bubble) {
    View trashContentView = getTrashContent();
    int trashCenterX = (trashContentView.getLeft() + (trashContentView.getMeasuredWidth() / 2));
    int trashCenterY = (trashContentView.getTop() + (trashContentView.getMeasuredHeight() / 2));
    int x = (trashCenterX - (bubble.getMeasuredWidth() / 2));
    int y = (trashCenterY - (bubble.getMeasuredHeight() / 2));
    bubble.getViewParams().x = x;
    bubble.getViewParams().y = y;
    windowManager.updateViewLayout(bubble, bubble.getViewParams());
  }

  private boolean checkIfBubbleIsOverTrash(Bubble bubble) {
    boolean result = false;
    if (trashView.getVisibility() == View.VISIBLE) {
      View trashContentView = getTrashContent();
      int trashWidth = trashContentView.getMeasuredWidth();
      int trashHeight = trashContentView.getMeasuredHeight();
      int trashLeft = (trashContentView.getLeft() - (trashWidth / 2));
      int trashRight = (trashContentView.getLeft() + trashWidth + (trashWidth / 2));
      int trashTop = (trashContentView.getTop() - (trashHeight / 2));
      int trashBottom = (trashContentView.getTop() + trashHeight + (trashHeight / 2));
      int bubbleWidth = bubble.getMeasuredWidth();
      int bubbleHeight = bubble.getMeasuredHeight();
      int bubbleLeft = bubble.getViewParams().x;
      int bubbleRight = bubbleLeft + bubbleWidth;
      int bubbleTop = bubble.getViewParams().y;
      int bubbleBottom = bubbleTop + bubbleHeight;
      if (bubbleLeft >= trashLeft && bubbleRight <= trashRight) {
        if (bubbleTop >= trashTop && bubbleBottom <= trashBottom) {
          result = true;
        }
      }
    }
    return result;
  }

  public void notifyBubbleRelease(Bubble bubble) {
    if (trashView != null) {
      if (checkIfBubbleIsOverTrash(bubble)) {
        bubblesService.removeBubble(bubble);
      }
      trashView.setVisibility(View.GONE);
    }
  }

  public static class Builder {
    private final Coordinator layoutCoordinator;

    public Builder(FloatingService service) {
      layoutCoordinator = getInstance();
      layoutCoordinator.bubblesService = service;
    }

    public Builder setTrashView(Trash trashView) {
      layoutCoordinator.trashView = trashView;
      return this;
    }

    public Builder setWindowManager(WindowManager windowManager) {
      layoutCoordinator.windowManager = windowManager;
      return this;
    }

    public Coordinator build() {
      return layoutCoordinator;
    }
  }

  private View getTrashContent() {
    return trashView.getChildAt(0);
  }
}

File: android/app/src/main/java/com/dismoi/scout/floating/layout/Coordinator.kt
package com.dismoi.scout.floating.layout

import android.view.View
import android.view.WindowManager
import com.dismoi.scout.floating.FloatingService

class Coordinator private constructor() {
  private var _trashView: Trash? = null
  private var windowManager: WindowManager? = null
  private var bubblesService: FloatingService? = null

  fun notifyBubblePositionChanged(bubble: Bubble, x: Int, y: Int) {
    if (_trashView != null) {
      _trashView!!.visibility = View.VISIBLE
      if (checkIfBubbleIsOverTrash(bubble)) {
        _trashView!!.applyMagnetism()
        _trashView!!.vibrate()
        applyTrashMagnetismToBubble(bubble)
      } else {
        _trashView!!.releaseMagnetism()
      }
    }
  }

  private fun applyTrashMagnetismToBubble(bubble: Bubble) {
    val trashContentView = trashContent
    val trashCenterX = trashContentView.left + trashContentView.measuredWidth / 2
    val trashCenterY = trashContentView.top + trashContentView.measuredHeight / 2
    val x = trashCenterX - bubble.measuredWidth / 2
    val y = trashCenterY - bubble.measuredHeight / 2
    bubble.viewParams!!.x = x
    bubble.viewParams!!.y = y
    windowManager!!.updateViewLayout(bubble, bubble.viewParams)
  }

  private fun checkIfBubbleIsOverTrash(bubble: Bubble): Boolean {
    var result = false
    if (_trashView!!.visibility == View.VISIBLE) {
      val trashContentView = trashContent
      val trashWidth = trashContentView.measuredWidth
      val trashHeight = trashContentView.measuredHeight
      val trashLeft = trashContentView.left - trashWidth / 2
      val trashRight = trashContentView.left + trashWidth + trashWidth / 2
      val trashTop = trashContentView.top - trashHeight / 2
      val trashBottom = trashContentView.top + trashHeight + trashHeight / 2
      val bubbleWidth = bubble.measuredWidth
      val bubbleHeight = bubble.measuredHeight
      val bubbleLeft = bubble.viewParams!!.x
      val bubbleRight = bubbleLeft + bubbleWidth
      val bubbleTop = bubble.viewParams!!.y
      val bubbleBottom = bubbleTop + bubbleHeight
      if (bubbleLeft >= trashLeft && bubbleRight <= trashRight) {
        if (bubbleTop >= trashTop && bubbleBottom <= trashBottom) {
          result = true
        }
      }
    }
    return result
  }

  fun notifyBubbleRelease(bubble: Bubble) {
    if (_trashView != null) {
      if (checkIfBubbleIsOverTrash(bubble)) {
        bubblesService!!.removeBubble(bubble)
      }
      _trashView!!.visibility = View.GONE
    }
  }

  class Builder(service: FloatingService?) {
    private val layoutCoordinator: Coordinator?
    fun setTrashView(trashView: Trash?): Builder {
      layoutCoordinator!!._trashView = trashView
      return this
    }

    fun setWindowManager(windowManager: WindowManager?): Builder {
      layoutCoordinator!!.windowManager = windowManager
      return this
    }

    fun build(): Coordinator? {
      return layoutCoordinator
    }

    init {
      layoutCoordinator = instance
      layoutCoordinator!!.bubblesService = service
    }
  }

  private val trashContent: View
    private get() = _trashView!!.getChildAt(0)

  companion object {
    private var INSTANCE: Coordinator? = null
    private val instance: Coordinator?
      private get() {
        if (INSTANCE == null) {
          INSTANCE = Coordinator()
        }
        return INSTANCE
      }
  }
}
